device.go

```diff
@@ -229,37 +231,174 @@ func virtioBlkDeviceHandler(_ context.Context, device pb.Device, spec *pb.Spec,
+	err := ioutil.WriteFile(overridePath, []byte("vfio-pci"), 0200)
+	if err != nil {
+		return fmt.Errorf("Failed to write %s: %s", overridePath, err)
+	}
+
+	driverPath := filepath.Join(sysPath, "driver")
+	driver, err := os.Readlink(driverPath)
+	if os.IsNotExist(err) {
+		// This just means the device isn't bound to any driver
+		driver = ""
+	} else if err != nil {
+		return fmt.Errorf("Failed to readlink %s: %s", driverPath, err)
+	}
+
+	fieldLogger.Debugf("VFIO: Device previously bound to '%s'", driver)
+
+	if driver != "vfio-pci" && driver != "" {
+		fieldLogger.Debugf("VFIO: Unbinding from %s", driver)
+		unbindPath := filepath.Join(sysPath, bdf, "driver", "unbind")
+		err = ioutil.WriteFile(unbindPath, []byte(bdf), 0200)
+		if err != nil {
+			return fmt.Errorf("Failed to write %s: %s:", unbindPath, err)
+		}
+	}
+
+	err = ioutil.WriteFile(pciProbeFile, []byte(bdf), 0200)
+	if err != nil {
+		return fmt.Errorf("Failed to write %s: %s", pciProbeFile, err)
+	}
+
+	fieldLogger.Info("VFIO: Rebind complete")
+	return nil
+}
+
+func vfioDeviceHandler(ctx context.Context, device pb.Device, spec *pb.Spec, s *sandbox, devIdx devIndex) error {
+	fieldLogger := agentLog.WithField("host-dev", device.ContainerPath)
+
+	var group string
+
+	for _, opt := range device.Options {
+		l := strings.Split(opt, "=")
+		if len(l) != 2 {
+			return fmt.Errorf("Malformed VFIO option \"%s\"", opt)
+		}
+
+		hostBdf := l[0]
+		guestPath := l[1]
+
+		guestAddr, err := getDevicePCIAddress(guestPath)
+		if err != nil {
+			return err
+		}
+
+		l = strings.Split(guestAddr, "/")
+		guestBdf := l[len(l)-1]
+
+		s.vfioDeviceMap[hostBdf] = guestBdf
+
+		err = vfioRebindDevice(guestBdf)
+		if err != nil {
+			return err
+		}
+
+		groupPath := filepath.Join(sysBusPrefix, guestBdf, "iommu_group")
+		groupLink, err := os.Readlink(groupPath)
+		if err != nil {
+			return fmt.Errorf("Failed to read %s: %s", groupPath, err)
+		}
+
+		if group == "" {
+			group = filepath.Base(groupLink)
+		} else if group != filepath.Base(groupLink) {
+			// If PCI devices associated with the same
+			// VFIO device (and therefore group) in the
+			// host don't end up in the same group in the
+			// guest, something has gone horribly wrong
+			return fmt.Errorf("%s is not in guest IOMMU group %s", guestBdf, group)
+		}
+		fieldLogger.WithField("host-bdf", hostBdf).WithField("guest-bdf", guestBdf).Debug("VFIO: Device complete")
+	}
+
+	vmpath, err := getDeviceName(s, filepath.Join("vfio", group))
+	if err != nil {
+		return fmt.Errorf("Couldn't find device for guest VFIO group %s: %s",
+			group, err)
+	}
+	device.VmPath = vmpath
+
+	fieldLogger.WithField("guest-dev", device.VmPath).Debug("VFIO: Group complete")
+
+	return updateSpecDevice(spec, devIdx, device.ContainerPath, device.VmPath, device.VmPath)
+}
+
+func VfioTranslateEnv(s *sandbox, env []string) []string {
+	newEnv := make([]string, 0, len(env))
+
+	for _, e := range env {
+		if strings.HasPrefix(e, "PCIDEVICE_") {
```

Comment:
I would make `PCIDEVICE_` configurable, although it currently seems pretty hard-coded in the rest of the stack. That function should probably take that as input. Ask around to see if this logic would not apply to other env variables.

Reply:
Hmm.  Configurable where, though?

Reply:
Just to make it clear: that would have to happen later, not in this PR. Just a note to do it later.

The agent has a `config.go` that lists some agent options. That would be a good starting point. I would make it a list of prefixes for the names of environment variables to rewrite, default to `[ "PCIDEVICE_" ]`.

Also, this is something that you might want to be able to override with annotations. See https://github.com/kata-containers/runtime/blob/master/virtcontainers/pkg/annotations/annotations.go#L233. I have not checked if agent annotations work well though.

Reply:
The environment variable is actually specific to the Kubernetes device plugin being used. For example the Intel SRIOV plugin inserts `PCIDEVICE_` (https://github.com/intel/sriov-network-device-plugin). Same with OpenShift SRIOV device plugin (https://github.com/openshift/sriov-network-device-plugin).
I don't think there is any standardization. 
Can this environment variable prefix be a runtime config in `configuration.toml` ?